src/swap/sideshift.js

```diff
@@ -0,0 +1,310 @@
+
+import { gt, lt, mul } from 'biggystring'
+import {
+  SwapAboveLimitError,
+  SwapBelowLimitError,
+  SwapPermissionError
+} from 'edge-core-js'
+import {
+  type EdgeCorePluginOptions,
+  type EdgeCurrencyWallet,
+  type EdgeFetchResponse,
+  type EdgeSpendInfo,
+  type EdgeSwapInfo,
+  type EdgeSwapPlugin,
+  type EdgeSwapQuote,
+  type EdgeSwapRequest,
+  type EdgeTransaction,
+  SwapCurrencyError
+} from 'edge-core-js/types'
+
+import { makeSwapPluginQuote } from '../swap-helpers.js'
+
+// Invalid currency codes should *not* have transcribed codes
+// because currency codes with transcribed versions are NOT invalid
+const CURRENCY_CODE_TRANSCRIPTION = {
+  // Edge currencyCode: exchangeCurrencyCode
+  USDT: 'usdtErc20'
+}
+
+const pluginId = 'sideshift'
+const swapInfo: EdgeSwapInfo = {
+  pluginId,
+  displayName: 'SideShift.ai',
+  supportEmail: 'help@sideshift.ai'
+}
+type FixedQuote = {
+  createdAt: string,
+  depositAmount: string,
+  depositMethod: string,
+  expiresAt: string,
+  id: string,
+  rate: string,
+  settleAmount: string,
+  settleMethod: string
+}
+
+type FixedQuoteRequestParams = {
+  depositMethod: string,
+  settleMethod: string,
+  depositAmount: string
+}
+
+type OrderRequest = {
+  createdAt: string,
+  createdAtISO: string,
+  expiresAt: string,
+  expiresAtISO: string,
+  depositAddress: {
+    address: string
+  },
+  depositMethod: string,
+  id: string,
+  orderId: string,
+  settleAddress: {
+    address: string
+  },
+  settleMethod: string,
+  depositMax: string,
+  depositMin: string,
+  quoteId: string,
+  settleAmount: string,
+  depositAmount: string,
+  deposits: Array<any>
+}
+
+type OrderRequestParams = {
+  type: string,
+  quoteId: string,
+  affiliateId: string,
+  sessionSecret?: string,
+  settleAddress: string
+}
+
+type Rate = {
+  rate: number,
+  min: string,
+  max: string,
+  error?: {
+    message: string
+  }
+}
+
+type Permission = {
+  createOrder: boolean,
+  createQuote: boolean
+}
+
+const dontUseLegacy = {}
+
+async function getAddress(
+  wallet: EdgeCurrencyWallet,
+  currencyCode: string
+): Promise<string> {
+  const addressInfo = await wallet.getReceiveAddress({ currencyCode })
+  return addressInfo.legacyAddress && !dontUseLegacy[currencyCode]
+    ? addressInfo.legacyAddress
+    : addressInfo.publicAddress
+}
+
+async function checkReplyForError(reply: EdgeFetchResponse): Promise<any> {
+  try {
+    return await reply.json()
+  } catch (e) {
+    throw new Error(`SideShift.ai returned error code ${reply.status}`)
+  }
+}
+
+async function checkRateForError(
+  rate: Rate,
+  request: EdgeSwapRequest,
+  depositAmount: string
+): Promise<any> {
+  const { fromCurrencyCode, toCurrencyCode, fromWallet, nativeAmount } = request
+  const { denominations, metaTokens } = fromWallet.currencyInfo
+  if (rate.error) {
+    throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+  }
+  const multiplier = denominations.find(d => d.name === fromCurrencyCode)
```

Comment:
@nilswx What do you think of this one?

Reply:
Could be `denominations.find(({ name }) => name === fromCurrencyCode)`, but it's more reading for me.

Unless you want to use lodash.

Reply:
I meant the repeated `.find` calls